### ISSUE TITLE
Grouped Window List: window thumbnails are hard to use because they disappear immediately when the pointer leaves them

### ISSUE THREAD

**Issue by LinuxOnTheDesktop** [linuxmint/cinnamon#8796]
```
 * Cinnamon version (cinnamon --version) 4.2.3
   - Please specify if you are using the daily builds PPA (https://launchpad.net/~linuxmint-daily-build-team/+archive/ubuntu/daily-builds). No.
 * Distribution - (Mint 17.2, Arch, Fedora 25, etc...) Mint 19.2
 * Graphics hardware *and* driver used: Intel UHD Graphics 620; Mesa
 * 32 or 64 bit: 64
 * Attach /home/<username>/.xsession-errors, or /var/log/syslog: not relevant
 ```

**Issue**

As soon as one's pointer goes just a tiny amount beyond a window thumbnail, the window thumbnail disappears (well, unless one is moving the pointer into another thumbnail). This makes the previews very hard to use - as [this video shows](https://app.box.com/s/nqhse33jtncl3053sgbm2t7eajxu35tf). Compare the behaviour on Windows 10, where the preview disappears only when the pointer has moved some (further) distance away from the thumbnail. [Here is a video](https://app.box.com/s/bi30wgs16xhml7gobwzjes84er1ktlz2) of the Windows behaviour. It seems to me that here Windows is very much getting things right.

**Steps to reproduce**

Please see the first of the two videos (the one showing Mint).

**Expected behaviour**

Please see the second of the two videos (the one showing Windows).

**Other information**

I looked at the applet's code to see whether I could make the required change myself, but it proved beyond me.

**Comment by LinuxOnTheDesktop** [comment]
Here are reproducible steps that allow one to appreciate the severity of the problem.

1. Have two windows from one application - e.g. from XFCE4-terminal - open.
2. Have Grouped Window List installed on a panel.
3. Use the aforementioned panel to try to switch to one particular window of the aforementioned application. That is: try to get one window rather than the other window focused.
4. If you try 3 a few times, then you will find the following. It is all too easy to end up with no windows focused, or end up with the wrong window from the application focused, or to end up with an application from a different application focused.

If you tire of my bug reports about usability, then please ask yourself following (even though I know that sometimes I have been overly keen and that sometimes I have made other mistakes in my bug reports). Should basic usability in your flagship DE really take a second place to, say, adding a feature to the xapps?

**Comment by smurphos** [comment]
This came up here https://github.com/linuxmint/cinnamon/issues/8741 in the context of themes that have a margin between the panel and the thumbnail and the user having difficulty moving the mouse quick enough to avoid the thumbnail disappearing. I use such a theme and don't personally find it a usability issue but the behaviour is controlled by the hard coded timeout I mentioned in the linked issue. 

Ideally this timeout should be exposed as user configurable setting.

**Comment by claudiux** [comment]
New option proposed in #8814.

**Comment by mtwebster** [comment]
I'm not sure I follow the issue here - is it the fact you're using 'hover peak' and wish to make the window appear when mousing over a particular thumbnail, then go into the work area and have enough time to focus that window?

Why not just click the thumbnail?  That will focus the window (unlike the old window list, where the thumbnails were just passive images).  I always felt the hover peak was just that, a peak at the window contents so you can decide if that's the one you want.

I couldn't really tell what you were trying to do in the windows video -it's hard to follow without the pointer (probably the built-in recorder?)

**Comment by LinuxOnTheDesktop** [comment]
@mtwebster 

I had not presented matters clearly enough. I am sorry for that. Let me better explain.

The problem is that GWL closes previews too readily. Here is how that problem manifests itself. Suppose that one of the icons on GWL is for an application that is running. In that case, when one moves the pointer to that icon - the icon, in GWL, for the running application - then one sees a preview. Now: if one moves the pointer from the icon to the preview, then the preview can disappear. That happens when one has moved the pointer to a location very slightly outside the area occupied by the preview. When one does that, the preview disappears, instantly. But one had wanted to focus that preview. A similar problem occurs if the app in question has multiple windows, generating multiple previews, and one puts the pointer just outside the whole preview area. I mean: if one sees previews A B C (all for one app), and one's pointer is just to the left of preview A, or just to the right of preview C, on indeed just above any of A, B or C, then the preview disappears and one fails to accomplish one's goal of focusing some particular window.

The short of it is: trying to focus a window via GWL is difficult, because GWL is too sensitive ('twitchy'): it will make a prevew - that one is trying to select - disappear much too readily. To summarise the summary: focusing a window via GWL doesn't work properly because the previews keep disappearing.

Smurphos may have diagnosed the problem in his post above. That said: changing Cinnamon theme did not seem to solve the problem (whereas the diagnosis offered by Smurphos, as I understood it, entailed that it should make a difference).

Incidentally: playing back the videos posted above revealed that the 'videos' application, if already running when told to play a video, did not the new video to the foreground (even when it, though running, was not actually playing a video already). Perhaps I should file a separate bug report for that.

**Comment by smurphos** [comment]
Hi,

Clem wasn't a fan of claudiux's PR to make the currently hardcoded GWL thumbnail display timeout of 50ms user configurable. The only reason that themes are relevant are in situations were the theme codes a margin (gap) between the panel and the thumbnail such that it is possible for the timeout to kick in and destroy the thumbnail as the user moves the mouse pointer (slowly) from the icon to the thumbnail.

Too long a default timeout does look weird if you move the mouse over several GWL icons with open instances, but maybe upping it from 50ms to somewhere in the region of 100ms to 200ms would be a good compromise.

**Comment by mtwebster** [comment]
Ok I understand - I found a theme that ridiculously magnifies part of the problem :)

I think making the close timeout larger, as mentioned, is the way to go, but it needs to be ignored, along with the open timeout, when an adjacent 'app group' is entered while the previous group's thumbnails are visible.

**Comment by mtwebster** [comment]
I opened a pr that address this, I'm honestly not sure if you need to be on the latest cinnamon or if you could just copy/paste the GWL applet only, but feel free to try :)

**Comment by LinuxOnTheDesktop** [comment]
I copied the new files to my computer and restarted the computer. GWL behaviour does seem improved (though it is possible that I think I detect a change when I do not).

However, (1) it seems to me that the applet is still a little bit too twitchy, i.e. that thumbnails still disappear a little (some 25%-40%?) too fast. 

Also, (2) it might be an idea to slow down the transition effect (by, again, 25-40%?) that occurs when
when one moves either from a GWL to a thumbnail (one of the small previews) or else from one thumbnail to another. I mean, the transition effect that is applied when _the full-screen_ preview appears.

Finally: (3) if the user moves the pointer from a thumbail to the full-screen preview, she might expect the previewed window to become focused. It doesn't. Perhaps it should.

**Comment by mtwebster** [comment]
I disagree with (3) - as I said in my first comments, I don't think that's the reason for the full-screen previews.  If this was the behavior, I would be unable to perform this action on a window that's, for instance, on a different monitor.  I think a click is necessary here, either on the panel item, or the thumbnail.  It reminds me a bit of the window setting we have - 'click' vs 'sloppy'.

As for the delay before disappearing, I'm actually concerned the current delay may be too much, and get in the way.  I think the only way to truly avoid the 'gap between panel and thumbnails' issue is a) prevent themes from doing this (while I think some of them look silly, this would be a last resort to me), or b) track pointer movement and basically created an invisible 'bridge' between panel item and thumbnail.

I can consider b down the road but not at the current stage of this release cycle - I'm already making more changes here than I'd normally be comfortable with.
